lib/src/types.dart

```diff
@@ -411,7 +411,15 @@ class TypedValue<T extends Object> {
   final Type<T> type;
   final T? value;
 
-  TypedValue(this.type, this.value);
+  /// Whether this value is deliberately describing a `NULL` value in SQL.
+  ///
+  /// For some types, in particular `json` and `jsonb`, the `null` value in Dart
+  /// can be mapped to a non-null SQL value (a textual or binary representation
+  /// of the `null` value in JSON, respectively).
+  final bool isSqlNull;
+
+  TypedValue(this.type, this.value, {bool? isSqlNull})
```

Comment:
From the original PR:

> Is there a case where `value != null` and we may set `isSqlNull: true`? That should be an argument error?

I agree that using `isSqlNull: true` for `value != null` should be an error.

> Maybe a `TypedValue.sqlNull(this.type)` constructor would be an explicit setter, and the above is always unset?

If we always assume it to be unset, we need to keep the current behavior and map `null` to `NULL`. We need a constructor to explicitly say "I want this Dart `null` to not translate into an SQL `NULL`", which `TypedValue.sqlNull(this.type)` can't do, it would just be a shortcut to the existing constructor.

Maybe we can do `TypedValue.nullValue(this.type, bool isNullInSql)`? That still feels kind of weird.

Reply:
Ah, you are right, it is the `isSqlNull: false` that we want to be filled there. It was just strange, because in my head the false should be the default value, and this makes it a bit confusing. Maybe inverting it to `isTypedNull` would be slightly closer to what I would find less confusing by the default value, however, I like the `isSqlNull` parameter name better.

I leave it up to you, if you are happier with this, let's go with it.

Reply:
I also like the symmetry with the names on the binding/result row sides. It's perhaps not the most obvious, but it's a feature most users won't have to care about and once they do, at least the concepts are the same everywhere in the package.

So I've added the error check but otherwise kept this as is.